### PR TITLE
Remove error for circular reference in members.

### PR DESCRIPTION
**Pull request by w-vi**
Keeps the error in inheritance references.
Tests are very basic, just checking that it does not produce an error,
the resulting ast is not validated in any way. Possible FIXME.

### CODE REVIEW

File: test/test-BlueprintParser.cc [hunk 1]
```diff
@@ -760,11 +760,11 @@ TEST_CASE("Report error when named sub type is referenced in nested members", "[
     ParseResult<Blueprint> blueprint;
     SectionParserHelper<Blueprint, BlueprintParser>::parse(source, BlueprintSectionType, blueprint, ExportSourcemapOption, Models(), &blueprint);
 
-    REQUIRE(blueprint.report.error.code == MSONError);
-    SourceMapHelper::check(blueprint.report.error.location, 35, 11);
+    REQUIRE(blueprint.report.error.code == Error::OK);
+    REQUIRE(blueprint.report.warnings.empty());
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Here we need to add a test to check that base type of `A` and `B` are `mson::ImplicitObjectBaseType`.

File: test/test-BlueprintParser.cc [hunk 2]
```diff
@@ -779,11 +779,11 @@ TEST_CASE("Report error when there are circular references in nested members", "
     ParseResult<Blueprint> blueprint;
     SectionParserHelper<Blueprint, BlueprintParser>::parse(source, BlueprintSectionType, blueprint, ExportSourcemapOption, Models(), &blueprint);
 
-    REQUIRE(blueprint.report.error.code == MSONError);
-    SourceMapHelper::check(blueprint.report.error.location, 50, 7);
+    REQUIRE(blueprint.report.error.code == Error::OK);
+    REQUIRE(blueprint.report.warnings.empty());
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Here we need to add `REQUIRE`s to check that base type of `A`, `B` and `C` are `mson::ImplicitObjectBaseType`.

File: test/test-BlueprintParser.cc [hunk 3]
```diff
@@ -796,11 +796,11 @@ TEST_CASE("Report error when named sub type is referenced in nested members when
     ParseResult<Blueprint> blueprint;
     SectionParserHelper<Blueprint, BlueprintParser>::parse(source, BlueprintSectionType, blueprint, ExportSourcemapOption, Models(), &blueprint);
 
-    REQUIRE(blueprint.report.error.code == MSONError);
-    SourceMapHelper::check(blueprint.report.error.location, 26, 11);
+    REQUIRE(blueprint.report.error.code == Error::OK);
+    REQUIRE(blueprint.report.warnings.empty());
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Here we need to add `REQUIRE`s to check that base type of `A` and `B` are `mson::ImplicitObjectBaseType`.

File: test/test-BlueprintParser.cc [hunk 4]
```diff
@@ -816,8 +816,8 @@ TEST_CASE("Report error when a resource attributes type is circularly referenced
     ParseResult<Blueprint> blueprint;
     SectionParserHelper<Blueprint, BlueprintParser>::parse(source, BlueprintSectionType, blueprint, ExportSourcemapOption, Models(), &blueprint);
 
-    REQUIRE(blueprint.report.error.code == MSONError);
-    SourceMapHelper::check(blueprint.report.error.location, 65, 13);
+    REQUIRE(blueprint.report.error.code == Error::OK);
+    REQUIRE(blueprint.report.warnings.empty());
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Here we need to add `REQUIRE`s to check that base type of `Post` and `B` are `mson::ImplicitObjectBaseType`.

File: test/test-BlueprintParser.cc [hunk 5]
```diff
@@ -832,6 +832,7 @@ TEST_CASE("Report error when named sub type is referenced as mixin", "[blueprint
     ParseResult<Blueprint> blueprint;
     SectionParserHelper<Blueprint, BlueprintParser>::parse(source, BlueprintSectionType, blueprint, ExportSourcemapOption, Models(), &blueprint);
 
+
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: What's with the extra line?

File: test/test-BlueprintParser.cc [hunk 6]
```diff
@@ -865,8 +884,8 @@ TEST_CASE("Report error when named type references itself in array", "[blueprint
     ParseResult<Blueprint> blueprint;
     SectionParserHelper<Blueprint, BlueprintParser>::parse(source, BlueprintSectionType, blueprint, ExportSourcemapOption, Models(), &blueprint);
 
-    REQUIRE(blueprint.report.error.code == MSONError);
-    SourceMapHelper::check(blueprint.report.error.location, 48, 26);
+    REQUIRE(blueprint.report.error.code == Error::OK);
+    REQUIRE(blueprint.report.warnings.empty());
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Here we need to add `REQUIRE`s to check that base type of `Comment` is `mson::ImplicitObjectBaseType`.

**Comment by pksunkara** [comment]
Unfortunately, when this error is removed, a few things happen when the parser continues. To make sure those things are correct, we need to add a few conditions to the tests which are changed currently. Please add them.

If you are confused about how to add them, you can see how it was done in the other tests and do it similarly.

**Comment by w-vi** [comment]
Cheers for tips on what to check, I had the feeling that it should go further than that, that's why I wrote possible FIXME.

**Comment by pksunkara** [comment]
@w-vi No problem. Just mention here when the PR is updated again.

**Comment by w-vi** [comment]
@pksunkara like this?

**Comment by pksunkara** [comment]
@w-vi Yup. Like that. :smile: Will review now.

File: test/test-BlueprintParser.cc [hunk 7]
```diff
@@ -760,11 +760,34 @@ TEST_CASE("Report error when named sub type is referenced in nested members", "[
...
+    REQUIRE(dsB.typeDefinition.baseType == mson::ImplicitObjectBaseType);
+    REQUIRE(dsB.typeDefinition.empty());
+    REQUIRE(dsB.sections.size() == 1);
+    REQUIRE(dsB.sections[0].klass == mson::TypeSection::MemberTypeClass);
+    REQUIRE(dsB.sections[0].content.elements().size() == 1);
+    REQUIRE(dsB.sections[0].content.elements().at(0).klass == mson::Element::PropertyClass);
+
+    mson::PropertyMember person = dsB.sections[0].content.elements().at(0).content.property;
+    REQUIRE(person.name.literal == "person");
+    REQUIRE(person.valueDefinition.typeDefinition.typeSpecification.name.symbol.literal == "A");
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Please check for person's `baseType` here.

File: test/test-BlueprintParser.cc [hunk 8]
```diff
@@ -779,11 +802,48 @@ TEST_CASE("Report error when there are circular references in nested members", "
...
+    REQUIRE(dsC.typeDefinition.baseType == mson::ImplicitObjectBaseType);
+    REQUIRE(dsC.typeDefinition.empty());
+    REQUIRE(dsC.sections.size() == 1);
+    REQUIRE(dsC.sections[0].klass == mson::TypeSection::MemberTypeClass);
+    REQUIRE(dsC.sections[0].content.elements().size() == 1);
+    REQUIRE(dsC.sections[0].content.elements().at(0).klass == mson::Element::PropertyClass);
+
+    mson::PropertyMember idC = dsC.sections[0].content.elements().at(0).content.property;
+    REQUIRE(idC.name.literal == "id");
+    REQUIRE(idC.valueDefinition.typeDefinition.typeSpecification.name.symbol.literal == "A");
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Please check for id's `baseType` here.

File: test/test-BlueprintParser.cc [hunk 9]
```diff
@@ -779,11 +802,48 @@ TEST_CASE("Report error when there are circular references in nested members", "
...
+    REQUIRE(dsA.typeDefinition.baseType == mson::ImplicitObjectBaseType);
+    REQUIRE(dsA.typeDefinition.empty());
+    REQUIRE(dsA.sections.size() == 1);
+    REQUIRE(dsA.sections[0].klass == mson::TypeSection::MemberTypeClass);
+    REQUIRE(dsA.sections[0].content.elements().size() == 1);
+    REQUIRE(dsA.sections[0].content.elements().at(0).klass == mson::Element::PropertyClass);
+
+    mson::PropertyMember idA = dsA.sections[0].content.elements().at(0).content.property;
+    REQUIRE(idA.name.literal == "id");
+    REQUIRE(idA.valueDefinition.typeDefinition.typeSpecification.name.symbol.literal == "B");
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Please check for id's `baseType` here.

File: test/test-BlueprintParser.cc [hunk 10]
```diff
@@ -796,11 +856,34 @@ TEST_CASE("Report error when named sub type is referenced in nested members when
...
+    REQUIRE(dsB.typeDefinition.baseType == mson::ImplicitObjectBaseType);
+    REQUIRE(dsB.typeDefinition.empty());
+    REQUIRE(dsB.sections.size() == 1);
+    REQUIRE(dsB.sections[0].klass == mson::TypeSection::MemberTypeClass);
+    REQUIRE(dsB.sections[0].content.elements().size() == 1);
+    REQUIRE(dsB.sections[0].content.elements().at(0).klass == mson::Element::PropertyClass);
+
+    mson::PropertyMember person = dsB.sections[0].content.elements().at(0).content.property;
+    REQUIRE(person.name.literal == "person");
+    REQUIRE(person.valueDefinition.typeDefinition.typeSpecification.name.symbol.literal == "A");
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Please check for person's `baseType` here.

File: test/test-BlueprintParser.cc [hunk 11]
```diff
@@ -816,8 +899,41 @@ TEST_CASE("Report error when a resource attributes type is circularly referenced
...
+    REQUIRE(dsB.typeDefinition.baseType == mson::ImplicitObjectBaseType);
+    REQUIRE(dsB.typeDefinition.empty());
+    REQUIRE(dsB.sections.size() == 1);
+    REQUIRE(dsB.sections[0].klass == mson::TypeSection::MemberTypeClass);
+    REQUIRE(dsB.sections[0].content.elements().size() == 1);
+    REQUIRE(dsB.sections[0].content.elements().at(0).klass == mson::Element::PropertyClass);
+
+    mson::PropertyMember posts = dsB.sections[0].content.elements().at(0).content.property;
+    REQUIRE(posts.name.literal == "posts");
+    REQUIRE(posts.valueDefinition.typeDefinition.typeSpecification.name.symbol.literal == "Post");
```
Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Please check for post's `baseType` here.

**Comment by pksunkara** [comment]
@w-vi Just a few more checks required and this PR should be good :smile:

**Comment by w-vi** [comment]
@pksunkara addressed, but travis seems to be very slow ...

**Pull request merged**